### ISSUE TITLE
Detached TypedArray semantics tests

### ISSUE THREAD

**Issue by littledan** [tc39/test262#368]
In ES2015, TypedArrays are supposed to throw on a number of operations (numeric property [[Get]] http://ecma-international.org/ecma-262/6.0/#sec-integerindexedelementget and [[Set]] http://ecma-international.org/ecma-262/6.0/#sec-integerindexedelementset as well as [[HasProperty]] http://ecma-international.org/ecma-262/6.0/#sec-integer-indexed-exotic-objects-hasproperty-p). While ES6 doesn't provide a way of neutering TypedArrays, each implementation does, and within the browser it is available via postMessage, so this could be provided in browser-specific harnesses. Currently, it seems that all browsers return undefined on [[Get]] from detached TypedArrays, even though that behavior is not spec'd anywhere (including the Khronos TypedArray spec, which indirectly references ES2015 for the behavior), and they ignore [[Set]] and [[HasProperty]].

**Comment by littledan** [comment]
@leobalter You proposed removing detached TypedArrays from the ES spec, but actually it'd be great if they were tested. This could be done with a change to the test harness.

**Comment by leobalter** [comment]
I agree it'd be great to have tests for detached TypedArrays and I'm afraid the current harness we have wouldn't provide a consistent solid case, as it could rely on external and non-ecma implementations.

I believe the best way to solve this problem is without relying on any implementation out of the scope provided by ECMA spec. How about proposing a `%TypedArray%.prototype.detach()` or even a `%TypedArray%.detach( typedArray )`?

There's no one better to discuss this but you. With one of these methods we could write tests to cover all the spec steps without any external resource.

I'm happy to use my open source time writing the steps for one of these proposed methods.

It won't be the first change on ecma262 to provide an easier way to provide better tests and we will end up with a better API, IMHO. I believe this method would have a similar purpose from what we have on Proxy's `revoke` method. They do different things but help covering internal steps.

**Comment by domenic** [comment]
This would have to be a new proposal that goes through the stages process. As it's useless for users, I wouldn't expect it to even make stage 0.

A better proposal is the existing transfer proposal, which seems to be stuck at stage 0.

**Comment by leobalter** [comment]
ArrayBuffer.transfer is on stage 1 for 11 months, maybe we could present this issue to give some priority to it.

**Comment by littledan** [comment]
I don't think we need to wait for putting this operation into ECMAScript (where I don't really see a great use case) to write the tests. We would just need to update the harness. I haven't looked into JSC or Chakra, but V8, SpiderMonkey and the web-based test harness all have straightforward ways to do this.

**Comment by leobalter** [comment]
We have a minimal coverage for operations wrt detached buffer on TypedArrays methods and internals. I still need to cover the constructors and static methods (from and of).

**Comment by leobalter** [comment]
At this point I believe we have a reasonable strategy and coverage for this.
